

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>PopGenStatistics Documentation &mdash; SNPio 1.1.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=859da57e" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=00f267c6"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example script" href="example_script.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SNPio
              <img src="_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="getting_started.html">Getting Started</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#importing-snpio">Importing SNPio</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#important-notes">Important Notes:</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#the-population-map-file">The Population Map File</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#reading-genotype-data">Reading Genotype Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#other-genotypedata-methods">Other GenotypeData Methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#filtering-genotype-data-with-nremover2">Filtering Genotype Data with NRemover2</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#genotypedata-properties">GenotypeData Properties</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#genotype-encoding-with-genotypeencoder">Genotype Encoding with GenotypeEncoder</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="getting_started.html#population-genetics-analysis-with-popgenstatistics">Population Genetics Analysis with PopGenStatistics</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">PopGenStatistics Documentation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PopGenStatistics"><code class="docutils literal notranslate"><span class="pre">PopGenStatistics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#key-features">Key Features</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li class="toctree-l4"><a class="reference internal" href="#methods-overview">Methods Overview</a></li>
<li class="toctree-l4"><a class="reference internal" href="#calculate-d-statistics">calculate_d_statistics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#detect-fst-outliers">detect_fst_outliers</a></li>
<li class="toctree-l4"><a class="reference internal" href="#summary-statistics">summary_statistics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#amova">amova</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#loading-and-parsing-phylogenetic-treeparser">Loading and Parsing Phylogenetic TreeParser</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#benchmarking-the-performance">Benchmarking the Performance</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#conclusion">Conclusion</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_script.html">Example script</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">snpio</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="getting_started.html">Getting Started</a></li>
      <li class="breadcrumb-item active">PopGenStatistics Documentation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/btmartin721/SNPio/blob/master/docs/source/pop_gen_statistics.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="popgenstatistics-documentation">
<h1>PopGenStatistics Documentation<a class="headerlink" href="#popgenstatistics-documentation" title="Link to this heading"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="PopGenStatistics">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">PopGenStatistics</span></span><a class="headerlink" href="#PopGenStatistics" title="Link to this definition"></a></dt>
<dd></dd></dl>

<p>The <cite>PopGenStatistics</cite> class is designed to perform a suite of population genetic analyses on SNP datasets, supporting methods to calculate D-statistics, Fst, heterozygosity, and more. This class is particularly useful for researchers studying population structure, diversity, and genetic variation within and between populations.</p>
<section id="key-features">
<h2>Key Features<a class="headerlink" href="#key-features" title="Link to this heading"></a></h2>
<blockquote>
<div><ul class="simple">
<li><p>Calculation of Patterson’s, partitioned, and D-foil D-statistics</p></li>
<li><p>Fst outlier detection using DBSCAN or bootstrapping</p></li>
<li><p>Calculation of heterozygosity, nucleotide diversity, and summary statistics</p></li>
<li><p>Perform Analysis of Molecular Variance (AMOVA) to assess genetic variation</p></li>
</ul>
</div></blockquote>
</section>
<section id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Link to this heading"></a></h2>
<p>PopGenStatistics is a part of the <cite>snpio</cite> package, which includes classes to calculate GenotypeEncoder, Plotting, and D-statistics, and depends on other packages such as <cite>numpy</cite>, <cite>pandas</cite>, <cite>scipy</cite>, <cite>sklearn</cite>, and <cite>statsmodels</cite>.</p>
<section id="quick-start">
<h3>Quick Start<a class="headerlink" href="#quick-start" title="Link to this heading"></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import necessary classes and initialize GenotypeData with your SNP data</span>
<span class="kn">from</span> <span class="nn">snpio</span> <span class="kn">import</span> <span class="n">VCFReader</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats</span> <span class="kn">import</span> <span class="n">PopGenStatistics</span>

<span class="c1"># Load SNP data and metadata into a GenotypeData object</span>
<span class="n">genotype_data</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">(</span>
    <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example_data/vcf_files/phylogen_subset14K_sorted.vcf.gz,</span>
    <span class="n">popmapfile</span><span class="o">=</span><span class="s2">&quot;example_data/popmaps/phylogen_nomx.popmap&quot;</span><span class="p">,</span>
    <span class="n">force_popmap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="c1"># Remove samples not in the popmap or vice versa.</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Initialize PopGenStatistics with genotype data</span>
<span class="n">pgs</span> <span class="o">=</span> <span class="n">PopGenStatistics</span><span class="p">(</span><span class="n">genotype_data</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><cite>calculate_d_statistics</cite>: Calculates D-statistics and saves them as a CSV. Also makes a plot of the D-statistics and returns a DataFrame with the statistics.</p></li>
<li><p><cite>detect_fst_outliers</cite>: Identifies Fst outliers between populations. Makes a plot of the Fst values and returns outlier SNPs as a DataFrame.</p></li>
<li><p><cite>observed_heterozygosity</cite>, <cite>expected_heterozygosity</cite>, <cite>nucleotide_diversity</cite>, and <cite>wrights_fst</cite>: Calculates core population genetic metrics.</p></li>
<li><p><cite>summary_statistics</cite>: Calculates and summarizes key metrics across populations. Makes informative plots and returns a dictionary of pandas DataFrame and Series objects with the results.</p></li>
<li><p><cite>amova</cite>: Conducts a hierarchical Analysis of Molecular Variance. Returns a dictionary with the AMOVA results. The algorithm supports parallel computation with bootstrapping to estimate variance and significance. The algorithm also supports a region map to group populations for AMOVA analysis and is adapted from Excoffier et al. (1992).</p></li>
<li><p><cite>neis_genetic_distance</cite>: Calculates Nei’s genetic distance between populations.</p></li>
<li><p><cite>plot_fst_heatmap</cite>: Plots a heatmap of pairwise Fst values between populations.</p></li>
<li><p><cite>plot_d_statistics</cite>: Plots D-statistics for Patterson’s D, partitioned D, and D-foil.</p></li>
<li><p><cite>plot_amova_results</cite>: Plots the AMOVA results.</p></li>
<li><p><cite>plot_summary_statistics</cite>: Plots summary statistics for heterozygosity, nucleotide diversity, and Fst.</p></li>
<li><p><cite>plot_neis_genetic_distance</cite>: Plots Nei’s genetic distance between populations.</p></li>
<li><p><cite>tajimas_d</cite>: Calculates Tajima’s D statistic for each population.</p></li>
<li><p><cite>plot_tajimas_d</cite>: Plots Tajima’s D statistic for each population.</p></li>
</ul>
<section id="core-methods">
<h3>Core Methods<a class="headerlink" href="#core-methods" title="Link to this heading"></a></h3>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Core Methods</span><a class="headerlink" href="#id1" title="Link to this table"></a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Method</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><cite>calculate_d_statistics</cite></p></td>
<td><p>Calculates D-statistics and saves them as a CSV. Also makes a plot of the D-statistics and returns a DataFrame with the statistics.</p></td>
</tr>
<tr class="row-odd"><td><p><cite>detect_fst_outliers</cite></p></td>
<td><p>Identifies Fst outliers between populations. Makes a plot of the Fst values and returns outlier SNPs as a DataFrame.</p></td>
</tr>
<tr class="row-even"><td><p><cite>observed_heterozygosity</cite></p></td>
<td><p>Calculates observed heterozygosity for each locus.</p></td>
</tr>
<tr class="row-odd"><td><p><cite>expected_heterozygosity</cite></p></td>
<td><p>Calculates expected heterozygosity for each locus.</p></td>
</tr>
<tr class="row-even"><td><p><cite>nucleotide_diversity</cite></p></td>
<td><p>Calculates nucleotide diversity for each locus.</p></td>
</tr>
<tr class="row-odd"><td><p><cite>summary_statistics</cite></p></td>
<td><p>Calculates and summarizes key metrics across populations. Makes informative plots and returns a dictionary of pandas DataFrame and Series objects with the results.</p></td>
</tr>
<tr class="row-even"><td><p><cite>amova</cite></p></td>
<td><p>Conducts a hierarchical Analysis of Molecular Variance. Returns a dictionary with the AMOVA results.</p></td>
</tr>
<tr class="row-odd"><td><p><cite>neis_genetic_distance</cite></p></td>
<td><p>Calculates Nei’s genetic distance between populations.</p></td>
</tr>
<tr class="row-even"><td><p><cite>tajimas_d</cite></p></td>
<td><p>Calculates Tajima’s D statistic for each population.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="calculate-d-statistics">
<h2>calculate_d_statistics<a class="headerlink" href="#calculate-d-statistics" title="Link to this heading"></a></h2>
<p>Calculates Patterson’s D-statistic, partitioned D-statistic, or D-foil for given populations. Bootstrap replicates and heterozygosity inclusion can be customized. The results are saved as a CSV and returned as a pandas DataFrame and a dictionary with mean overall results, and a plot of the D-statistics is generated.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">,</span> <span class="n">stats_summary</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">calculate_d_statistics</span><span class="p">(</span>
    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;patterson&quot;</span><span class="p">,</span>
    <span class="n">population1</span><span class="o">=</span><span class="s2">&quot;EA&quot;</span><span class="p">,</span>
    <span class="n">population2</span><span class="o">=</span><span class="s2">&quot;GU&quot;</span><span class="p">,</span>
    <span class="n">population3</span><span class="o">=</span><span class="s2">&quot;TT&quot;</span><span class="p">,</span>
    <span class="n">outgroup</span><span class="o">=</span><span class="s2">&quot;OG&quot;</span><span class="p">,</span>
    <span class="n">num_bootstraps</span><span class="o">=</span><span class="mi">1000</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="detect-fst-outliers">
<h2>detect_fst_outliers<a class="headerlink" href="#detect-fst-outliers" title="Link to this heading"></a></h2>
<p>Detects Fst outliers using bootstrapping or DBSCAN. The method returns Fst outlier SNPs along with their associated population pairs.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">outliers</span><span class="p">,</span> <span class="n">pvals_df</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">detect_fst_outliers</span><span class="p">(</span>
    <span class="n">correction_method</span><span class="o">=</span><span class="s2">&quot;bonf&quot;</span><span class="p">,</span> <span class="c1"># perform Bonferroni P-value adjustments.</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="c1"># significance level after P-value adjustment.</span>
    <span class="n">use_bootstrap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">1000</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="summary-statistics">
<h2>summary_statistics<a class="headerlink" href="#summary-statistics" title="Link to this heading"></a></h2>
<p>Calculates a comprehensive suite of summary statistics, including heterozygosity, nucleotide diversity, and Fst. Results can be plotted or returned as a dictionary.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">summary</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">summary_statistics</span><span class="p">()</span>

<span class="c1"># Access overall summary statistics</span>
<span class="n">ho_overall</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;overall&quot;</span><span class="p">][</span><span class="s2">&quot;Ho&quot;</span><span class="p">]</span>
<span class="n">he_overall</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;overall&quot;</span><span class="p">][</span><span class="s2">&quot;He&quot;</span><span class="p">]</span>
<span class="n">pi_overall</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;overall&quot;</span><span class="p">][</span><span class="s2">&quot;Pi&quot;</span><span class="p">]</span>

<span class="c1"># Access population-specific summary statistics</span>
<span class="n">ho_pops</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;populations&quot;</span><span class="p">][</span><span class="s2">&quot;Ho&quot;</span><span class="p">]</span>
<span class="n">he_pops</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;populations&quot;</span><span class="p">][</span><span class="s2">&quot;He&quot;</span><span class="p">]</span>
<span class="n">pi_pops</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;populations&quot;</span><span class="p">][</span><span class="s2">&quot;Pi&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>The per-population summary statistics are stored in a dictionary with population labels as keys and pandas DataFrames as values.</p>
</section>
<section id="amova">
<h2>amova<a class="headerlink" href="#amova" title="Link to this heading"></a></h2>
<p>Conducts an Analysis of Molecular Variance (AMOVA) to assess genetic variation within and among populations. The algorithm supports parallel computation with bootstrapping to estimate variance and significance. The method also supports a region map to group populations for AMOVA analysis. It is adapted from Excoffier et al. (1992).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Map populations to regions for AMOVA analysis.</span>
<span class="n">regionmap</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;EA&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GU&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TT&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TC&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ON&quot;</span><span class="p">:</span> <span class="s2">&quot;Ornate&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DS&quot;</span><span class="p">:</span> <span class="s2">&quot;Ornate&quot;</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">amova_results</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">amova</span><span class="p">(</span>
    <span class="n">regionmap</span><span class="o">=</span><span class="n">regionmap</span><span class="p">,</span> <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">amova_results</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>{
    &quot;Among_region_variance&quot;: 0.123,
    &quot;Among_population_within_region_variance&quot;: 0.456,
    &quot;Within_population_variance&quot;: 0.789,
    &quot;Phi_RT&quot;: 0.123, # Phi component among regions.
    &quot;Phi_PR&quot;: 0.456, # Phi component among populations within regions.
    &quot;Phi_PT&quot;: 0.789, # Phi component within populations.
    &quot;Phi_PT_p_value&quot;: 0.012, # Significance of Phi_PT.
    &quot;Phi_PR_p_value&quot;: 0.345, # Significance of Phi_PR.
    &quot;Phi_RT_p_value&quot;: 0.678, # Significance of Phi_RT.
}
</pre></div>
</div>
<section id="advanced-usage">
<h3>Advanced Usage<a class="headerlink" href="#advanced-usage" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p><strong>Bootstrap Replicates in Fst Calculation</strong>: To estimate the variance of Fst across SNPs, use the <cite>detect_fst_outliers</cite> method with <cite>use_bootstrap=True</cite>.</p></li>
<li><p><strong>Multiple Population Comparisons in D-statistics</strong>: The <cite>calculate_d_statistics</cite> method supports extended D-statistics calculations (e.g., D-foil) across more than four populations.</p></li>
<li><p><strong>Plotting</strong>: By default, plots for each metric are generated and saved. Customize <cite>plot_kwargs</cite> within your <cite>GenotypeData</cite> object if specific styling or debug configurations are needed.</p></li>
</ul>
</section>
<section id="additional-information">
<h3>Additional Information<a class="headerlink" href="#additional-information" title="Link to this heading"></a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>SNP data must be encoded in a compatible format.</p></li>
<li><p><cite>genotype_data.popmap</cite> must map samples to population labels accurately.</p></li>
<li><p>It is advised to run the Fst, D-statistic, and AMOVA calculations with sufficient bootstraps to obtain statistically robust estimates.</p></li>
<li><dl class="simple">
<dt><strong>Parallelization</strong>:</dt><dd><ul>
<li><p>Many methods support parallel computation by specifying <cite>n_jobs=-1</cite> to use all available CPU cores, optimizing for large SNP datasets.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Random Seed</strong>:</dt><dd><ul>
<li><p>For reproducibility, set a random seed in the <cite>amova</cite> method to ensure consistent results across runs.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Region Map</strong>:</dt><dd><ul>
<li><p>The <cite>amova</cite> method supports a region map to group populations for AMOVA analysis. Ensure the region map is correctly defined to avoid errors.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Outlier Detection</strong>:</dt><dd><ul>
<li><p>The <cite>detect_fst_outliers</cite> method supports outlier detection using bootstrapping or DBSCAN. Adjust the <cite>correction_method</cite> and <cite>alpha</cite> parameters to control the significance level.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Summary Statistics</strong>:</dt><dd><ul>
<li><p>The <cite>summary_statistics</cite> method calculates and summarizes key metrics across populations. Customize the <cite>save_plots</cite> parameter to save plots to disk.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<p>We hope this documentation helps you get started with the <cite>PopGenStatistics</cite> class in the <cite>snpio</cite> package. For more information, please refer to the <cite>snpio</cite> documentation or contact the developers for assistance.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="getting_started.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="example_script.html" class="btn btn-neutral float-right" title="Example script" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>